### PR TITLE
ARROW-10209: [Python] Support positional options in compute functions

### PR DESCRIPTION
**Pull request by pitrou**
This makes compute functions easier to use, for example here the required "pattern" option doesn't need to be passed by name:
```
>>> pc.split_pattern("abacab", "a")
<pyarrow.ListScalar: ['', 'b', 'c', 'b']>
```

... and producing the following doc at the prompt:
```
split_pattern(strings, /, pattern, *, max_splits=-1, reverse=False, options=None, memory_pool=None)
    Split string according to separator.

    Split each string according to the exact `pattern` defined in
    SplitPatternOptions.  The output for each string input is a list
    of strings.

    The maximum number of splits and direction of splitting
    (forward, reverse) can optionally be defined in SplitPatternOptions.

    Parameters
    ----------
    strings : Array-like or scalar-like
        Argument to compute function
    pattern : optional
        Parameter for SplitPatternOptions constructor. Either `options`
        or `pattern` can be passed, but not both at the same time.
    max_splits : optional
        Parameter for SplitPatternOptions constructor. Either `options`
        or `max_splits` can be passed, but not both at the same time.
    reverse : optional
        Parameter for SplitPatternOptions constructor. Either `options`
        or `reverse` can be passed, but not both at the same time.
    options : pyarrow.compute.SplitPatternOptions, optional
        Parameters altering compute function semantics.
    memory_pool : pyarrow.MemoryPool, optional
        If not passed, will allocate memory from the default memory pool.
```

### CODE REVIEW

**Comment by pitrou** [comment]
Based on #11951.

File: python/pyarrow/tests/test_compute.py [hunk 1]
```diff
@@ -1854,6 +1896,11 @@ def test_count():
     assert pc.count(arr, mode='only_valid').as_py() == 3
     assert pc.count(arr, mode='only_null').as_py() == 2
     assert pc.count(arr, mode='all').as_py() == 5
+    assert pc.count(arr, 'all').as_py() == 5
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_compute.py:1899]: I am wondering here, should we rather make the `mode` keyword keyword-only now we are changing this? 
I don't find it super clear what the "all" in `pc.count(arr, 'all')` would mean.

**Review by jorisvandenbossche — COMMENTED** [review]
Nice! Added some small comments

File: python/pyarrow/tests/test_compute.py (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/tests/test_compute.py:1899]: Actually, it's not that "mode" is that a clarifying keyword name either .., so ignore my comment.

File: python/pyarrow/compute.py [hunk 2]
```diff
@@ -326,243 +335,6 @@ def cast(arr, target_type, safe=True):
...
-    Examples
-    --------
-    >>> import pyarrow as pa
-    >>> import pyarrow.compute as pc
-    >>> arr = pa.array([1, 1, 2, 2, 3, 2, 2, 2])
-    >>> modes = pc.mode(arr, 2)
-    >>> modes[0]
-    <pyarrow.StructScalar: {'mode': 2, 'count': 5}>
-    >>> modes[1]
-    <pyarrow.StructScalar: {'mode': 1, 'count': 2}>
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/compute.py:516]: Indeed there is. I'm not sure how to do that in a maintainable way, though. Should we have a separate file with function stubs and docstrings?

File: python/pyarrow/compute.py [hunk 3]
```diff
@@ -326,243 +335,6 @@ def cast(arr, target_type, safe=True):
...
-    """
-    options = ModeOptions(n, skip_nulls=skip_nulls, min_count=min_count)
-    return call_function("mode", [array], options)
-
-
-def filter(data, mask, null_selection_behavior='drop'):
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/compute.py:522]: Hmm. I don't really like the idea of maintaining these special case wrappers, though. Or perhaps we could raise a deprecation warning before removing them.

@amol- Thoughts?

File: python/pyarrow/tests/test_compute.py (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/tests/test_compute.py:1899]: Yes, allowing positional arguments is always a balancing act...

File: python/pyarrow/compute.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/compute.py:516]: For example a separate file `_compute_docstrings.py` with contents such as:
```python

function_examples = {
    "mode": """
        >>> import pyarrow as pa
        >>> import pyarrow.compute as pc
        >>> arr = pa.array([1, 1, 2, 2, 3, 2, 2, 2])
        >>> modes = pc.mode(arr, 2)
        >>> modes[0]
        <pyarrow.StructScalar: {'mode': 2, 'count': 5}>
        >>> modes[1]
        <pyarrow.StructScalar: {'mode': 1, 'count': 2}>
        """,
}
```

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/compute.py:516]: Yes, something like that is what I had in mind. And then we can add those automatically to the generated docstrings.

[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/compute.py:516]: Ok, done. Can you take a look again?

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/compute.py:516]: Looks good!

File: python/pyarrow/compute.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply python/pyarrow/compute.py:522]: Let's maybe just change this.

**Review by jorisvandenbossche — APPROVED** [review]

**Pull request closed without merging**